### ISSUE TITLE
Device dehydration v2

### ISSUE THREAD

**Issue by BillCarsonFr** [vector-im/element-meta#922]
## User story

As a user, if I log out my last device, other users should still be able to send messages to me and I should be able to read them once I come back and log in with some device.

## Concept

Initial [device dehydration feature](https://github.com/matrix-org/matrix-doc/pull/2697) has several issues:
- Tricky to implement, as it requires that the device rehydration be done before any other API calls
- User loses the advantage of verifying their other devices via emoji or QR code (passphrase needed anyhow to rehydrate)

It's now replaced with MSC3814.

### Groundwork
- [ ] https://github.com/vector-im/element-meta/issues/1898
- [x] Rust crypto SDK https://github.com/matrix-org/matrix-rust-sdk/pull/2349
- [ ] https://github.com/vector-im/crypto-internal/issues/126
- [x] MSC updates/Polishing https://github.com/element-hq/element-meta/issues/2602
- [x] Synapse support part 2 https://github.com/matrix-org/synapse/pull/16010
- [ ] https://github.com/matrix-org/synapse/pull/16068

### Initial Dehydrated device support
- [x] IOS SDK https://github.com/matrix-org/matrix-ios-sdk/pull/1807
- [x] iOS App https://github.com/vector-im/element-ios/pull/7630
- [ ] https://github.com/element-hq/element-meta/issues/2278
- [ ] https://github.com/element-hq/element-meta/issues/2601
- [ ] https://github.com/matrix-org/matrix-rust-sdk/issues/4313
- [ ] https://github.com/matrix-org/matrix-rust-sdk/issues/4466
- [ ] https://github.com/element-hq/element-meta/issues/2603

### Production ready on Web
- [ ] https://github.com/element-hq/element-web/issues/29387
- [ ] https://github.com/element-hq/element-meta/issues/2703
- [ ] https://github.com/element-hq/element-meta/issues/2753
- [ ] https://github.com/element-hq/element-meta/issues/2704
- [ ] https://github.com/element-hq/element-meta/issues/2799
- [ ] https://github.com/element-hq/element-web/issues/29084
- [ ] https://github.com/element-hq/element-web/issues/29081
- [ ] https://github.com/element-hq/element-web/issues/29080
- [ ] https://github.com/element-hq/element-web/issues/29264
- [ ] https://github.com/element-hq/element-web/issues/28760
- [ ] https://github.com/element-hq/element-web/issues/29579
- [ ] https://github.com/element-hq/element-meta/issues/2828

### Future work
- [ ] https://github.com/vector-im/element-meta/issues/1978
- [ ] Deprecate and remove legacy device dehydration (MSC2697)
- [ ] https://github.com/element-hq/element-meta/issues/2635
- [ ] EX full support for dehydrated device v2
- [ ] https://github.com/element-hq/crypto-internal/issues/402

**Comment by pmaier1** [comment]
> Design/UX inputs needed for settings and option to enable disable

As discussed today, we want to make this feature transparent to the user as far as possible. As it requires key backup being enabled, we'd bind the availability of dehydrated devices to the setting to enable/disable key backup.

**Comment by emmaburton1** [comment]
Work on this can be timesheeted to the project on https://element-io.atlassian.net/browse/PSB-87

**Comment by poljar** [comment]
So a status update here:
Ruma PR: https://github.com/ruma/ruma/pull/1605
Rust SDK branch: https://github.com/matrix-org/matrix-rust-sdk/tree/poljar/device-dehydration
Synapse branch that can be used for testing: https://github.com/poljar/synapse/tree/poljar/msc3814-v2

@stefanceriu above are the main ingredients that you'll need to get started on this on the iOS side, the Rust SDK branch depends on the Ruma PR, so you only need the later two.

**Comment by poljar** [comment]
> Work on this can be timesheeted to the project on https://element-io.atlassian.net/browse/PSB-87

Where exactly do I find this in the myhours thingy? I can't seem to find a PSB-87 in the task list.

**Comment by pmaier1** [comment]
> Where exactly do I find this in the myhours thingy? I can't seem to find a PSB-87 under the BWI task list.

It should just be the title of this ticket.

**Comment by emmaburton1** [comment]
@poljar I've added the task to relevant customer retainer project.

**Comment by pmaier1** [comment]
Considering that device dehydration
- will first be a labs feature and we probably can't ask all the end users to turn it on over and over again (when logging out and logging in again)
- the server-side support will also be an experimental feature at first

can we automatically enable the feature on the client if the server-side setting is enabled? Probably doesn't make sense to have two switches to enable it and this way we could control availability centrally.

**Comment by t3chguy** [comment]
@pmaier1 we don't tend to have HS-driven labs features due to the potential for data loss or work preventing bugs whilst thinking you are running a stable release of Element

**Comment by pmaier1** [comment]
then we need a separate switch on the backend to force enabling the labs feature on the client. otherwise this will be unusable.

**Comment by t3chguy** [comment]
Element web/desktop already support Such via config.json but not via anything HS driven.

**Comment by Xyz00777** [comment]
hi everyone, what is the state, because in https://github.com/element-hq/element-web/pull/28173 dehydration devices are removed...

**Comment by ara4n** [comment]
i believe that's the v1 code being removed to make room for the v2 code.

**Comment by uhoreg** [comment]
For "production ready" on Element Web, we should have (aside from the current tasks in progress):

- [the SSSS issues resolved](https://github.com/element-hq/crypto-internal/issues/257)
- [gossiping of the Dehydration Key](https://github.com/element-hq/element-meta/issues/2703)
- better way of enabling the feature (currently, requires a flag in `.well-known` -- maybe not needed, if we're just going to turn on decryption everywhere)
- [any required UI improvements](https://github.com/element-hq/element-meta/issues/278)

I'd also *really* like us make the API changes (https://github.com/matrix-org/matrix-spec-proposals/pull/3814/files#r948469437 and https://github.com/matrix-org/matrix-spec-proposals/pull/3814/files#r1540896531) and have the MSC ready for FCP.  This isn't technically required for production ready from a product perspective, but it would be very sad if had dehydrated devices "production ready", but have the MSC stall out.

**Comment by mxandreas** [comment]
Trying to summarize the current state from a user perspective.

* The user must have key storage & recovery must be enabled.
  * This is partially a technical dependency as the dehydration key (used to encrypt the offline device before its creation/upload to the server) needs to be stored on the server.
  * However, also, if user goes offline (no logged in devices), the only way to verify a new device next time they log in, is with a recovery key.
* The user needs to log out and log in again from Element Web (EW) after the feature has been enabled on the server.
  * This login/logout is not considered a significant friction as the feature is meant for users who have to log out regularly from all of their devices. This issue will be addressed.
  * Example: Users log out at the end of the workday. Admin enables the feature on Monday. Tuesday morning users still have messages that fail to decrypt. However, starting from Wednesday morning, all messages sent over night will decrypt successfully.
  * The enabling is done in Synapse with `msc3814_enabled` as this is currently a beta/experimental feature. There is currently no user-specific toggle to enable/disable this as it is not sure if this is needed.
* Only EW is currently supported, this means:
  * If the user uses only EW, all messages sent while they were offline should be always successfully received and decrypted when they log in again.
  * If the user does not use EW at all, the messages sent while the user was offline can not be decrypted.
  * In case of mixed usage (e.g. EW and EX):
    * User needs to logout/login with EW at least once for the feature to take effect (as already explained above).
    * If user logs into EX after they were offline - the messages sent while the user was offline can not be decrypted (on EX).
    * As soon as the user logs into EW - the messages sent while the user was offline are successfully decrypted, and after that also get decrypted on EX.
 * The user has to verify their device with recovery key in order to decrypt the messages that were sent while they were offline.
   * This is FYI only. If the user can verify with another device, it means that they were not offline and hence they did not have the problem in the first place.
 * The user is expected to be not offline for extended periods (e.g. several months) as their offline device has limited capacity to fit all the keys required to decrypt the messages that are sent while they are offline.
   * The exact limits - e.g. how many messages can the offline device fit - are currently not precisely measured.
 * If the user is online for extended periods (e.g. several months), their offline device *may* exhaust its capacity.
   * This happens only when all of the user's current devices were verified with another device. As there is no device that was verified with recovery key, the offline device can not be rotated (its capacity freed).
   * This problem will be addressed by gossiping dehydration key between devices so any device can rotate the offline device.

**Comment by uhoreg** [comment]
> * This login/logout is not considered a significant friction as the feature is meant for users who have to log out regularly from all of their devices.

Users to log in and out regularly is one of the use cases for dehydrated devices.  But it is also useful for people who only have one device, and then lose that device.  So the requirement to log in may be OK for now, but it definitely should be fixed eventually.

> * In case of mixed usage (e.g. EW and EX): 
>   * ...
>   * As soon as the user logs into EW - the messages sent while the user was offline are successfully decrypted, and after that also get decrypted on EX.

Assuming that they have key backup enabled, so that EX can obtain the keys that were received by EW

**Comment by ara4n** [comment]
I poked at this a few weeks ago in order to get confidence we could unblock it on EW - before I lose them (given I didn't open PRs), the findings were:

- Don't delete the dehydrated device at logout when MAS syncs the device list (i.e. unbreak dehydration on EW)
    - https://github.com/element-hq/synapse/compare/develop...matthew/mas-sync-devices-preserve-dehydrated
- Spot when the dehydration key is missing and give a toast to recover it (https://github.com/element-hq/element-web/issues/29579)
    - https://github.com/matrix-org/matrix-js-sdk/compare/develop...matthew/dehydration-detect-missing-key
    - https://github.com/element-hq/element-web/compare/develop...matthew/dehydration-recover-prompt
- Spot when the dehydration key gets changed by another device on the account (https://github.com/element-hq/element-web/issues/28760)
    - https://github.com/matrix-org/matrix-js-sdk/compare/develop...matthew/dehydration-reconcile-on-rotate
- Gossip the dehydration key when you verify a device (https://github.com/element-hq/element-meta/issues/2703)
    - https://github.com/matrix-org/matrix-rust-sdk/compare/main...matthew/dehydration-gossip-on-verify
- Gossip the dehydration key when it changes (https://github.com/element-hq/element-meta/issues/2753)
    - https://github.com/matrix-org/matrix-rust-sdk/compare/matthew/dehydration-gossip-on-verify...matthew/dehydration-gossip-on-change
- Fix an unrelated circular import problem in EW which was making the tests flake
    - https://github.com/element-hq/element-web/compare/develop...matthew/fix-readywatchingstore-circular-import

Switching from POST to GET for /events as per https://github.com/element-hq/element-meta/issues/2704

- https://github.com/uhoreg/matrix-doc/compare/shrivelled\_sessions...matrix-org:matrix-spec-proposals:matthew/dehydrated-device-events-get
- https://github.com/element-hq/synapse/compare/matthew/dehydrated-device-events-get
- https://github.com/matrix-org/matrix-js-sdk/compare/matthew/dehydrated-device-events-get

Some of these have now been taken over by the team, but some haven't, hence recording them here lest we forget them

---

**Update 2026-08-24**: went back through these branches to check which were still valid vs overtaken by upstream work, and opened draft PRs (rebased and re-tested against current develop/main) for the ones still standing:

Superseded by the team (no PR needed):
- `synapse matthew/mas-sync-devices-preserve-dehydrated` → landed as https://github.com/element-hq/synapse/pull/19892
- POST→GET for `/events`:
  - MSC3814 itself was updated by @uhoreg (POST→GET plus `from` pagination), replacing `matthew/dehydrated-device-events-get` on the spec repo
  - `synapse matthew/dehydrated-device-events-get` → landed as https://github.com/element-hq/synapse/pull/19896 (+ https://github.com/element-hq/synapse/pull/19897 for pagination conventions)
  - `js-sdk matthew/dehydrated-device-events-get` → landed as https://github.com/matrix-org/matrix-js-sdk/pull/5392 (+ https://github.com/matrix-org/matrix-js-sdk/pull/5397)

Still relevant, now up as draft PRs:
- Detect a missing dehydration key (https://github.com/element-hq/element-web/issues/29579):
  - https://github.com/matrix-org/matrix-js-sdk/pull/5492 (`isDehydratedDeviceKeyMissing()`)
  - https://github.com/element-hq/element-web/pull/34797 (recovery toast; blocked on the js-sdk PR)
- Reconcile the dehydration key with 4S before rotating (https://github.com/element-hq/element-web/issues/28760):
  - https://github.com/matrix-org/matrix-js-sdk/pull/5493
- Gossip the dehydration key on verify (https://github.com/element-hq/element-meta/issues/2703):
  - https://github.com/matrix-org/matrix-rust-sdk/pull/6909
- Gossip the dehydration key when it changes (https://github.com/element-hq/element-meta/issues/2753):
  - https://github.com/matrix-org/matrix-rust-sdk/pull/6910 (stacked on 6909)
- ReadyWatchingStore circular-import fix:
  - https://github.com/element-hq/element-web/pull/34796
